### PR TITLE
Fix: flickering issue if Doc is modified outside N++

### PR DESCRIPTION
**Pull request by SinghRajenM**
Fixes issue: #2010 , #1018, #1761 
@Ede123 Thank for giving idea about [fix](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/2010#issuecomment-282321687).

### CODE REVIEW

File: PowerEditor/src/Notepad_plus.cpp [hunk 1]
```diff
@@ -5001,6 +5001,18 @@ void Notepad_plus::notifyBufferChanged(Buffer * buffer, int mask)
...
+					// Idea is :
+					//		1. First get the current caret position.
+					//		2. Send WM_LBUTTONDOWN and WM_LBUTTONUP message (which acts as double click at 0,0)
+					//		3. Now set caret back to original position.
+					//
+					auto curPos = _pEditView->execute(SCI_GETCURRENTPOS);
+					::PostMessage(_pEditView->getHSelf(), WM_LBUTTONDOWN, 0, 0);
+					::PostMessage(_pEditView->getHSelf(), WM_LBUTTONUP, 0, 0);
+					::PostMessage(_pEditView->getHSelf(), SCI_SETSEL, curPos, curPos);
+					// Now the flickering issue is not seen.
```
Comment (dail8859) [inline PowerEditor/src/Notepad_plus.cpp]: Would a single call to `_pEditView->execute(SCI_GRABFOCUS);` work?

Comment (SinghRajenM) [inline PowerEditor/src/Notepad_plus.cpp]: Sorry, it does not work. :(

Comment (SinghRajenM) [inline PowerEditor/src/Notepad_plus.cpp]: I mean _pEditView->execute(SCI_GRABFOCUS); does not work

Comment (dail8859) [inline PowerEditor/src/Notepad_plus.cpp]: Good to know. It was just a thought.

**Comment by Ede123** [comment]
You should probably check what messages are sent as a result of clicking the text area.
I guess sending one single message could already be enough to sort this out (or maybe only changing order of messages somewhere, as I assume this is everything the conflicting commit did)

**Comment by SinghRajenM** [comment]
Sending WM_LBUTTONDOWN itself is enough to handle the flickering, but then caret behavior is unpredictable.
Sometimes, caret is set at top while sometime it is set somewhere else. 

Therefore, send WM_LBUTTONDOWN (and obviously WM_LBUTTONDOWN should be released so send WM_LBUTTONUP, other scintilla will consider that user has pressed down, so start selecting text based in mouse move).
Now we don't know where is the caret, so we must set it to original position. Similar thing is done few lines down.

``` C++
if (didDialog)
		{
			auto curPos = _pEditView->execute(SCI_GETCURRENTPOS);
			::PostMessage(_pEditView->getHSelf(), WM_LBUTTONUP, 0, 0);
			::PostMessage(_pEditView->getHSelf(), SCI_SETSEL, curPos, curPos);
			if (::IsIconic(_pPublicInterface->getHSelf()))
				::ShowWindow(_pPublicInterface->getHSelf(), SW_RESTORE);

			if (doCloseDoc) // buffer has been deleted, cannot (and no need to) go on
				return;
		}
```

**Comment by Ede123** [comment]
I just created #2956 which contains and improves upon this fix.
I assume you won't mind if Don commited that one instead @SinghRajenM ?

**Comment by Ede123** [comment]
@donho Have you seen my comment? Should I rebase #2956?

**Pull request closed without merging**